File: backend/schemas/lab.py
from datetime import datetime
from typing import Optional, List, Dict, Any
from pydantic import BaseModel

# Lab模式
class LabBase(BaseModel):
    title: str
    description: str
    category: str
    difficulty: str
    points: int
    docker_image: str
    port_mapping: str
    hints: Optional[List[str]] = None
    resource_limits: Optional[Dict[str, Any]] = None

class LabCreate(LabBase):
    flag: str

class LabUpdate(LabBase):
    flag: Optional[str] = None
    is_active: Optional[bool] = None

class LabResponse(LabBase):
    id: int
    flag: Optional[str] = None
    is_active: bool
    created_at: datetime
    updated_at: datetime

    class Config:
        orm_mode = True

# LabInstance模式
class LabInstanceBase(BaseModel):
    lab_id: int
    user_id: int
    container_id: str
    instance_url: Optional[str] = None
    status: str

class LabInstanceCreate(LabInstanceBase):
    pass

class LabInstanceUpdate(BaseModel):
    status: Optional[str] = None
    end_time: Optional[datetime] = None

class LabInstanceResponse(LabInstanceBase):
    id: int
    start_time: datetime
    end_time: Optional[datetime] = None

    class Config:
        orm_mode = True

# LabProgress模式
class LabProgressBase(BaseModel):
    lab_id: int
    user_id: int
    status: str
    score: int = 0
    attempts: int = 0

class LabProgressCreate(LabProgressBase):
    pass

class LabProgressUpdate(BaseModel):
    status: Optional[str] = None
    score: Optional[int] = None
    attempts: Optional[int] = None
    completed_at: Optional[datetime] = None

class LabProgressResponse(LabProgressBase):
    id: int
    completed_at: Optional[datetime] = None
    created_at: datetime
    updated_at: datetime

    class Config:
        orm_mode = True

# 统计数据模式
class UserStats(BaseModel):
    completedCount: int
    totalPoints: int
    totalHours: int
    rank: int

class SkillRadarData(BaseModel):
    categories: List[str]
    scores: List[float]

class ProgressTrendData(BaseModel):
    dates: List[str]
    counts: List[int]

# Flag提交模式
class FlagSubmission(BaseModel):
    flag: str

class FlagResponse(BaseModel):
    correct: bool
    message: str 

File: frontend/src/api/lab.ts
import request from '@/utils/request'
import type { AxiosResponse } from 'axios'

// API 响应类型定义
interface Lab {
  id: number
  title: string
  description: string
  category: string
  difficulty: 'easy' | 'medium' | 'hard'
  points: number
  status: 'not_started' | 'created' | 'running' | 'stopped' | 'completed' | 'error'
  completionRate: number
  solvedCount: number
  created_at: string | null
  hints: string[]
  docker_image: string
  port_mapping: string
  resource_limits: {
    cpu: string
    memory: string
  }
}

interface UserStats {
  totalHours: number
  completedCount: number
  points: number
  rank: number
}

interface SkillRadarData {
  categories: string[]
  scores: number[]
}

interface ProgressTrendData {
  dates: string[]
  counts: number[]
}

interface LabStatus {
  status: 'not_started' | 'created' | 'running' | 'stopped' | 'completed' | 'error'
  instance_url?: string
}

interface LabResponse {
  status: string
  message: string
  instance_id?: number
}

interface FlagSubmission {
  correct: boolean
  message: string
}

// API 请求函数
export const labApi = {
  // 获取靶场列表
  getList(): Promise<AxiosResponse<Lab[]>> {
    return request({
      url: '/api/practice/labs',
      method: 'get'
    })
  },

  // 获取靶场详情
  getLabDetail(id: number): Promise<AxiosResponse<Lab>> {
    return request({
      url: `/api/practice/labs/${id}`,
      method: 'get'
    })
  },

  // 启动靶场
  startLab(id: number): Promise<AxiosResponse<LabResponse>> {
    return request({
      url: `/api/practice/labs/${id}/start`,
      method: 'post'
    })
  },

  // 停止靶场
  stopLab(id: number): Promise<AxiosResponse<LabResponse>> {
    return request({
      url: `/api/practice/labs/${id}/stop`,
      method: 'post'
    })
  },

  // 提交Flag
  submit(id: number, flag: string): Promise<AxiosResponse<FlagSubmission>> {
    return request({
      url: `/api/practice/labs/${id}/submit`,
      method: 'post',
      data: { flag }
    })
  },

  // 获取靶场状态
  getStatus(id: number): Promise<AxiosResponse<LabStatus>> {
    return request({
      url: `/api/practice/labs/${id}/status`,
      method: 'get'
    })
  },

  // 获取用户统计数据
  getUserStats(): Promise<AxiosResponse<UserStats>> {
    return request({
      url: '/api/practice/stats',
      method: 'get'
    })
  },

  // 获取技能雷达图数据
  getSkillRadarData(): Promise<AxiosResponse<SkillRadarData>> {
    return request({
      url: '/api/practice/stats/skill-radar',
      method: 'get'
    })
  },

  // 获取进度趋势数据
  getProgressTrend(days: number = 30): Promise<AxiosResponse<ProgressTrendData>> {
    return request({
      url: '/api/practice/stats/progress-trend',
      method: 'get',
      params: { days }
    })
  }
} 